org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.java

```diff
@@ -0,0 +1,107 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2019 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.concurrent;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.eclipse.xtext.service.OperationCanceledManager;
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.xbase.lib.Functions.Function0;
+import org.eclipse.xtext.xbase.lib.Functions.Function1;
+import org.eclipse.xtext.xbase.lib.Functions.Function2;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * @author kosyakov - Initial contribution and API
+ * @since 2.11
+ */
+@Singleton
+public class RequestManager {
+
+	@Inject
+	private ExecutorService parallel;
+
+	@Inject
+	private OperationCanceledManager operationCanceledManager;
+
+	private final ExecutorService queue = Executors.newSingleThreadExecutor(
+			new ThreadFactoryBuilder().setDaemon(true).setNameFormat("RequestManager-Queue-%d").build());
+
+	private List<AbstractRequest<?>> requests = new ArrayList<>();
+
+	/**
+	 * An orderly shutdown of this request manager.
+	 */
+	public void shutdown() {
+		queue.shutdown();
+		parallel.shutdown();
+		cancel();
+	}
+
+	/**
+	 * Run the given cancellable logic as a read request.
+	 */
+	public <V> CompletableFuture<V> runRead(Function1<? super CancelIndicator, ? extends V> cancellable) {
+		return submit(new ReadRequest<>(this, cancellable, parallel));
+	}
+
+	/**
+	 * Perform the given write and run the cancellable logic afterwards.
+	 */
+	public <U, V> CompletableFuture<V> runWrite(
+			Function0<? extends U> nonCancellable,
+			Function2<? super CancelIndicator, ? super U, ? extends V> cancellable) {
```

Comment:
Can't we use `Supplier` and `BiFunction` here now?

Reply:
I'd opt for `Callable` and some `FunctionWithExceptionLike` thing where both allow exceptions to be thrown.
I've to think about the impact on backwards compatibility, though and if this is a thing that we should keep compatible.